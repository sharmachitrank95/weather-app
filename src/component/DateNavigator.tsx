import React from 'react';
import {List,ListItem, Grid} from '@material-ui/core'
interface IDateNavigatorProps{
    dates:any,
    currentDate:any,
    handleDateChange:(e:any,value:any)=>void
}

function getListItemStyle(match:boolean) {
    if(match){
        return {
            backgroundColor: 'grey',
        };
    }
}
export const DateNavigator:React.SFC<IDateNavigatorProps> =(props) => {
        let dates;
        dates = props.dates.map((date:any) => {
               return <ListItem 
               button 
               component={"a"} 
               value={date} 
               onClick={e => props.handleDateChange(e,date)}
               style={getListItemStyle(date === props.currentDate)}>
               {date}
               </ListItem>
           })
        
        return(
            <Grid item xs={2}>
                <h4 style={{paddingLeft:'1rem'}}>Choose Date:</h4>
                <List component={"nav"}>
                    {dates}
                </List>
            </Grid>
        )
}

export default DateNavigator;