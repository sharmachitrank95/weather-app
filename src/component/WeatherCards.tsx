import React from 'react';
import {Card, CardHeader, CardContent, Typography, Grid} from '@material-ui/core';

interface IWeatherCardsProps{
    WeatherCardData:[]
}
interface IWeatherCardsStates{

}

export class WeatherCards extends React.Component<IWeatherCardsProps,IWeatherCardsStates>{
    constructor(props:any) {
        super(props);
    }

    render(){
        let cards;
        cards = this.props.WeatherCardData.map((cardData:any)=>{
            return (
                <Grid item xs={"auto"}>
                    <Card>
                        <CardContent>
                            <Typography>
                               Time: {cardData.dt_txt.split(' ')[1]}
                            </Typography>
                            <Typography variant="h3">
                                {cardData.weather[0].description}
                            </Typography>
                            <Typography>
                                Max Temp: {cardData.main.temp_max}
                            </Typography>
                            <Typography>
                                Min Temp: {cardData.main.temp_min}
                            </Typography>
                            <Typography>
                                Pressure: {cardData.main.pressure}
                            </Typography>
                            <Typography>
                                Humidity: {cardData.main.humidity}
                            </Typography>
                        </CardContent>
                    </Card>
                </Grid>
            )
        })
        return(
            <Grid item xs={10}>
                <Grid container spacing={24} style={{flexGrow:1}}>
                   {cards}
                </Grid>
            </Grid>
        )
        
    }
}

export default WeatherCards;