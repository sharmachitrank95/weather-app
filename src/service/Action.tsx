
export const loginData = (userName:string) =>
{
    return{
    type: 'LOGIN',
    payload: {userName: userName}
    }
}

export const logoutData = () =>
{
    return{
    type: 'LOGOUT',
    payload: {username: ''}
    }
}