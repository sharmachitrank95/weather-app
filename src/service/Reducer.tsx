const initialState = {
    userName:'',
    password:''
  }
export default (state = initialState, action:any) => {
    switch (action.type) {
      case 'LOGIN':{
            localStorage.setItem('userName',action.payload.userName);
            console.log("welcome"+ action.userName);
            return Object.assign({}, state, {
                userName: action.payload.userName
            });
            break;
        }
      case 'LOGOUT':{
            localStorage.setItem('userName','');
            console.log("Bye Bye");
            return Object.assign({}, state, {
                userName: action.payload.username
            });
            break;
        }
      default:
        return Object.assign({}, state, {
            userName: localStorage.getItem('userName'),
        });
    }
  }