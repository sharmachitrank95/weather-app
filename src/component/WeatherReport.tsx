import React from 'react';
import {DateNavigator} from '../component/DateNavigator';
import {WeatherCards} from '../component/WeatherCards';
import * as _ from 'lodash';
import { Grid } from '@material-ui/core';
import CircularProgress from '@material-ui/core/CircularProgress';

interface IWeatherReportProps{
    weatherData:any
}
interface IWeatherReportStates{
    date:any,
    cardsData:any,
    currentDate:any
}

export class WeatherReport extends React.Component<IWeatherReportProps,IWeatherReportStates>{
    public availableDates:Array<string>;
    constructor(props:any) {
        super(props);
        this.state = {cardsData:null,date:null,currentDate:null};
        this.availableDates = [];
        this.handleDateChange = this.handleDateChange.bind(this);
        this.setCardData = this.setCardData.bind(this);
        this.fetchAvailableDates();
    }

    fetchAvailableDates(){
        this.props.weatherData.list.map((element:any) => {
            element = element.dt_txt.split(' ')[0];
            if(this.availableDates.indexOf(element) === -1){
                this.availableDates.push(element);
            }
        })
    }
    componentWillMount(){
        this.setCardData(this.availableDates[0],this.props.weatherData);
    }

    handleDateChange(e:any,value:any){
       this.setCardData(value,this.props.weatherData);
    }

    setCardData(date:any,weatherData:any){
        let data = _.filter(weatherData.list,(element:any) =>{
           return element.dt_txt.split(' ')[0] === date;
        })
        this.setState({
            currentDate: date,
            cardsData:_.cloneDeep(data)
        })
    }

    render(){
        return(
            <Grid container spacing={8}>
                <DateNavigator dates={this.availableDates} handleDateChange={this.handleDateChange} currentDate={this.state.currentDate}/>
               {!!this.state.cardsData && <WeatherCards WeatherCardData={this.state.cardsData}/>}
            </Grid>
        )
    }
}

export default WeatherReport;

