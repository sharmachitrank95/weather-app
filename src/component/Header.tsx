import React from 'react';
import {logoutData} from '../service/Action'
import { Grid, Button, AppBar, Toolbar, Link } from '@material-ui/core';

interface IHeaderProps {
    store:any,
    cityName:string,
    setCurrentLocation:(e:any) => void
}

interface IHeaderStates {
    userName:string
}

export class Header extends React.Component<IHeaderProps, IHeaderStates> {
    constructor(props: any) {
        super(props)
        this.onSubmitClick = this.onSubmitClick.bind(this);
        this.state = {userName: this.props.store.getState().userName}
    }

    onSubmitClick(e:any){
        this.props.store.dispatch(logoutData());
        location.reload();
    }

    render() {
        return (
                <Grid container alignItems={"center"} style={{marginBottom:'1rem'}}>
                    <Grid item style={{flexGrow:1}}>
                        <AppBar position={"static"} style={{flexGrow:1, flexDirection:'row'}}>
                            <Toolbar style={{flexGrow:1, flexDirection:'row'}}>
                                <Grid item id="welcomeUser" style={{flexGrow:1}} > Welcome {this.state.userName}</Grid>
                                <Grid item>{this.props.cityName}</Grid>
                                <Link style={{color:"white", cursor:"pointer", padding:'1rem'}} onClick={this.props.setCurrentLocation}><i>choose current location</i></Link>
                                <Grid item ><Button onClick={this.onSubmitClick} variant={"contained"} >Logout</Button></Grid>
                            </Toolbar>
                        </AppBar>
                    </Grid>
                </Grid> 
        )
    }
}

export default Header;;