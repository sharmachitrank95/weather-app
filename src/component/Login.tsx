import React from 'react';
import { Redirect } from 'react-router';
import {loginData,logoutData} from '../service/Action'
import { Grid, Input, Button } from '@material-ui/core';

interface ILoginProps {
    store:any,
}

interface ILoginStates {
    userName:string,
    password:string
}

function getHomeStyle() {
    return {
        padding:"1rem",
    }
}

export class Login extends React.Component<ILoginProps, ILoginStates> {
    constructor(props: any) {
        super(props)
        this.state = {userName:'',password:''};
        this.onSubmitClick = this.onSubmitClick.bind(this);
        this.handleNameChange = this.handleNameChange.bind(this);
        this.handlePasswordChange = this.handlePasswordChange.bind(this);
        this.renderRedirect = this.renderRedirect.bind(this);
    }

    renderRedirect(){
        if (!!localStorage.getItem('userName') && localStorage.getItem('userName')!.length > 0) {
          return <Redirect to={process.env.PUBLIC_URL +'/webApp'} />
        }
    }

    onSubmitClick(e:any){
        this.props.store.dispatch(loginData(this.state.userName));
        this.renderRedirect();
        location.reload();
    }

    handleNameChange(e:any){
        this.setState({
            userName: e.target.value
        });
    }

    handlePasswordChange(e:any){
        this.setState({
            password:e.target.value
        });
    }

    render() {
        return (
            <Grid container direction={"column"} alignItems={"center"} justify={"center"} style={{height:"100%"}}>
                {this.renderRedirect()}
                <Grid item style={getHomeStyle()} >
                    <Input id="userName" placeholder="name" required  autoComplete={'false'} type="name" onChange={this.handleNameChange}/>
                </Grid> 
                 <Grid item style={getHomeStyle()}>  
                    <Input placeholder="password"  type="password" />
                </Grid>
                <Grid item style={getHomeStyle()}>  
                    <Button id="loginButton" variant={"contained"} disabled = {this.state.userName.length == 0} onClick={this.onSubmitClick}>LogIn</Button>
                </Grid>
            </Grid>
        )
    }
}

export default Login;