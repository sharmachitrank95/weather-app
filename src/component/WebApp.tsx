import React from 'react';
import { Redirect } from 'react-router';
import {Header} from '../component/Header';
import {WeatherReport} from '../component/WeatherReport';
import * as _ from 'lodash';
import { CircularProgress } from '@material-ui/core';

interface IWebAppStates{
    xCoordinate: Number,
    yCoordinate:Number,
    weatherData:any,
    isBusy:boolean
}
interface IWebAppProps{
    store:any
}

export class WebApp extends React.Component<IWebAppProps, IWebAppStates>{
    private cityName:string = '';
    constructor(props:any){
        super(props);
        this.state = {xCoordinate:30.734301,yCoordinate:76.793297,weatherData:null,isBusy:false};
        this.setCurrentLocation = this.setCurrentLocation.bind(this);
        this.renderRedirect = this.renderRedirect.bind(this);
    }

    setCurrentLocation(){
        navigator.geolocation.getCurrentPosition((position)=>{
            if(position){
                this.setState({
                    xCoordinate:position.coords.latitude,
                    yCoordinate:position.coords.longitude
                })
            }
            this.getData();
        });
    }

    shouldComponentUpdate(nextProps:any,nextState:any){
        if(!this.state.weatherData || (!!this.state.weatherData && this.state.weatherData.city.name !== nextState.weatherData.city.name)){
            return true;
        }
        if(this.state.isBusy !== nextState.isBusy){
            return true;
        }
        return false;
    }
    
    componentDidMount(){
        this.getData();
    }

    getData(){
        this.setState({
             isBusy:true
        })
        fetch('https://cors-anywhere.herokuapp.com/api.openweathermap.org/data/2.5/forecast?lat='+ this.state.xCoordinate +'&lon='+ this.state.yCoordinate +'&appid=0cefc283fa0c3ad68584e2c30b643c94',
        {
            headers:{'Content-Type': 'text/plain','Access-Type':'application/json'},
        })
        .then((data:any)=>{
           return data.json();
        })
        .then((data)=>{
            this.cityName = data.city.name;
            this.setState({
                isBusy:false,
                weatherData:_.cloneDeep(data) 
            })
        })
    }
    renderRedirect(){
        if (!localStorage.getItem('userName') || localStorage.getItem('userName')!.length == 0) {
          return <Redirect to={process.env.PUBLIC_URL +'/login'} />
        }
    }

    render(){
        let weatherReport;
        let spinner;
        weatherReport = <div id="weatherReport">
            <Header store={this.props.store} cityName={this.cityName} setCurrentLocation={this.setCurrentLocation}/>
            {!!this.state.weatherData && <WeatherReport weatherData={this.state.weatherData}/>}
        </div>
        spinner = <CircularProgress style={{position:'fixed',top:'50%',right:'50%',left:'50%',bottom:'50%'}} size={100}></CircularProgress>
        return(
            <React.Fragment>
                {this.renderRedirect()}
                {this.state.isBusy? spinner: weatherReport}
            </React.Fragment> 
        )
    }
}

export default WebApp;
