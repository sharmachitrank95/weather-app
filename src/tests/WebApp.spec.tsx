import { configure, shallow, mount } from 'enzyme';
import {WebApp} from '../component/WebApp'
import {createStore} from 'redux';
import myApp from '../service/Reducer';
import * as React from 'react';
import Adapter from 'enzyme-react-adapter-future';
import { loginData, logoutData } from '../service/Action';

configure({ adapter: new Adapter() });

const store = createStore(myApp);

describe('<WebApp />',()=>{
    it('should render correctly',()=>{
        const WebAppComponent = <WebApp store={store} />;
        const wrapper = shallow(WebAppComponent);
        expect(wrapper).toMatchSnapshot();   
    });

    it('should fetch data',()=>{
        store.dispatch(loginData("Test"));

        const mockSuccessResponse = {};
        const mockJsonPromise = Promise.resolve(mockSuccessResponse); // 2
        const mockFetchPromise = Promise.resolve({ // 3
        json: () => mockJsonPromise,
        });
        jest.spyOn(window,'fetch').mockImplementation(() => mockFetchPromise); // 4
        
        const wrapper = shallow(<WebApp store={store} />); // 5
                                
        expect(window.fetch).toHaveBeenCalledTimes(1);
    })

    it('should render data correctly',()=>{
        store.dispatch(loginData("Test"));

        const mockSuccessResponse = {"cod":"200","message":0.0045,"cnt":39,"list":[{"dt":1551063600,"main":{"temp":287.35,"temp_min":285.171,"temp_max":287.35,"pressure":1012.8,"sea_level":1012.8,"grnd_level":966.75,"humidity":73,"temp_kf":2.18},"weather":[{"id":803,"main":"Clouds","description":"broken clouds","icon":"04d"}],"clouds":{"all":80},"wind":{"speed":2.86,"deg":110.502},"sys":{"pod":"d"},"dt_txt":"2019-02-25 03:00:00"},{"dt":1551139200,"main":{"temp":282.483,"temp_min":282.483,"temp_max":282.483,"pressure":1012.51,"sea_level":1012.51,"grnd_level":966.1,"humidity":92,"temp_kf":0},"weather":[{"id":500,"main":"Rain","description":"light rain","icon":"10n"}],"clouds":{"all":32},"wind":{"speed":4,"deg":114.007},"rain":{"3h":0.7},"sys":{"pod":"n"},"dt_txt":"2019-02-26 00:00:00"},{"dt":1551225600,"main":{"temp":282.264,"temp_min":282.264,"temp_max":282.264,"pressure":1008.45,"sea_level":1008.45,"grnd_level":962.2,"humidity":96,"temp_kf":0},"weather":[{"id":501,"main":"Rain","description":"moderate rain","icon":"10n"}],"clouds":{"all":92},"wind":{"speed":2.6,"deg":43.0042},"rain":{"3h":7.37},"sys":{"pod":"n"},"dt_txt":"2019-02-27 00:00:00"},{"dt":1551236400,"main":{"temp":281.905,"temp_min":281.905,"temp_max":281.905,"pressure":1009.82,"sea_level":1009.82,"grnd_level":963.47,"humidity":98,"temp_kf":0},"weather":[{"id":501,"main":"Rain","description":"moderate rain","icon":"10d"}],"clouds":{"all":92},"wind":{"speed":3.11,"deg":59.0013},"rain":{"3h":4.38},"sys":{"pod":"d"},"dt_txt":"2019-02-27 03:00:00"},{"dt":1551322800,"main":{"temp":280.508,"temp_min":280.508,"temp_max":280.508,"pressure":1017.46,"sea_level":1017.46,"grnd_level":970.7,"humidity":84,"temp_kf":0},"weather":[{"id":800,"main":"Clear","description":"clear sky","icon":"01d"}],"clouds":{"all":0},"wind":{"speed":2.21,"deg":349.002},"rain":{},"sys":{"pod":"d"},"dt_txt":"2019-02-28 03:00:00"},{"dt":1551398400,"main":{"temp":280.258,"temp_min":280.258,"temp_max":280.258,"pressure":1016.53,"sea_level":1016.53,"grnd_level":969.6,"humidity":78,"temp_kf":0},"weather":[{"id":800,"main":"Clear","description":"clear sky","icon":"01n"}],"clouds":{"all":0},"wind":{"speed":3.3,"deg":333.001},"rain":{},"sys":{"pod":"n"},"dt_txt":"2019-03-01 00:00:00"}],"city":{"id":1274746,"name":"Chandigarh","coord":{"lat":30.7343,"lon":76.7933},"country":"IN","population":914371}};
        const mockJsonPromise = Promise.resolve(mockSuccessResponse); 
        const mockFetchPromise = Promise.resolve({ 
        json: () => mockJsonPromise,
        });
        jest.spyOn(window,'fetch').mockImplementation(() => mockFetchPromise); 
        
        const wrapper = shallow(<WebApp store={store} />); 
                                
        expect(window.fetch).toHaveBeenCalled();
    })
    afterEach(() =>{
        store.dispatch(logoutData());  
    })
})