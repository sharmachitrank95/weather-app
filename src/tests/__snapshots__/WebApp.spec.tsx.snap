// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WebApp /> should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WebApp
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Redirect
          from="/"
          push={false}
          to="/login"
        />,
        <WithStyles(CircularProgress)
          size={100}
          style={
            Object {
              "bottom": "50%",
              "left": "50%",
              "position": "fixed",
              "right": "50%",
              "top": "50%",
            }
          }
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "from": "/",
          "push": false,
          "to": "/login",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "size": 100,
          "style": Object {
            "bottom": "50%",
            "left": "50%",
            "position": "fixed",
            "right": "50%",
            "top": "50%",
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Redirect
            from="/"
            push={false}
            to="/login"
          />,
          <WithStyles(CircularProgress)
            size={100}
            style={
              Object {
                "bottom": "50%",
                "left": "50%",
                "position": "fixed",
                "right": "50%",
                "top": "50%",
              }
            }
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "from": "/",
            "push": false,
            "to": "/login",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "size": 100,
            "style": Object {
              "bottom": "50%",
              "left": "50%",
              "position": "fixed",
              "right": "50%",
              "top": "50%",
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
